ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Node/ProtocolVersions/Internal.hs

```diff
@@ -0,0 +1,187 @@
+import           Ouroboros.Consensus.Byron.Ledger.Block (ByronBlock)
+import           Ouroboros.Consensus.Byron.Node (ProtocolParams (..))
+import           Ouroboros.Consensus.Cardano.Block (CardanoBlock, CardanoEras)
+import           Ouroboros.Consensus.Protocol.Praos (Praos)
+import           Ouroboros.Consensus.Protocol.TPraos (TPraos)
+import           Ouroboros.Consensus.Shelley.Eras (AllegraEra, AlonzoEra,
+                     BabbageEra, ConwayEra, MaryEra, ShelleyEra)
+import           Ouroboros.Consensus.Shelley.Ledger.Block (ShelleyBlock)
+import           Ouroboros.Consensus.Shelley.Node (ProtocolParams (..))
+
+{-------------------------------------------------------------------------------
+  Cardano protocol versions
+-------------------------------------------------------------------------------}
+
+newtype CardanoProtocolVersions c = CardanoProtocolVersions_ {
+    getCardanoProtocolVersions :: PerEraProtocolVersion (CardanoEras c)
+  }
+
+{-# COMPLETE CardanoProtocolVersions #-}
+pattern CardanoProtocolVersions ::
+     ProtocolVersion ByronBlock
+  -> ProtocolVersion (ShelleyBlock (TPraos c) (ShelleyEra c))
+  -> ProtocolVersion (ShelleyBlock (TPraos c) (AllegraEra c))
+  -> ProtocolVersion (ShelleyBlock (TPraos c) (MaryEra    c))
+  -> ProtocolVersion (ShelleyBlock (TPraos c) (AlonzoEra  c))
+  -> ProtocolVersion (ShelleyBlock (Praos  c) (BabbageEra c))
+  -> ProtocolVersion (ShelleyBlock (Praos  c) (ConwayEra  c))
+  -> CardanoProtocolVersions c
+pattern CardanoProtocolVersions {
+      protVerByron
+    , protVerShelley
+    , protVerAllegra
+    , protVerMary
+    , protVerAlonzo
+    , protVerBabbage
+    , protVerConway
+    } =
+  CardanoProtocolVersions_ {
+      getCardanoProtocolVersions = PerEraProtocolVersion {
+          getPerEraProtocolVersion =
+               WrapProtocolVersion protVerByron
+            :* WrapProtocolVersion protVerShelley
+            :* WrapProtocolVersion protVerAllegra
+            :* WrapProtocolVersion protVerMary
+            :* WrapProtocolVersion protVerAlonzo
+            :* WrapProtocolVersion protVerBabbage
+            :* WrapProtocolVersion protVerConway
+            :* Nil
+        }
+    }
+
+cardanoMainnetProtocolVersions :: Bool -> CardanoProtocolVersions c
+cardanoMainnetProtocolVersions experimentalHardForksEnabled =
+    CardanoProtocolVersions {
+        protVerByron   = Byron.Update.ProtocolVersion 3 0 0
+      , protVerShelley = SL.ProtVer (SL.natVersion @3) 0
+      , protVerAllegra = SL.ProtVer (SL.natVersion @4) 0
+      , protVerMary    = SL.ProtVer (SL.natVersion @5) 0
+      , protVerAlonzo  = SL.ProtVer (SL.natVersion @7) 0
+      , protVerBabbage = if experimentalHardForksEnabled then
+                           SL.ProtVer (SL.natVersion @9) 0
+                         else
+                           SL.ProtVer (SL.natVersion @8) 0
+      , protVerConway  = SL.ProtVer (SL.natVersion @9) 0
+      }
```

Comment:
What about adding some documentation here what these numbers mean and where they are coming from?

One step could be to define these using the relevant ledger functions:
```suggestion
cardanoMainnetProtocolVersions :: Bool -> CardanoProtocolVersions c
cardanoMainnetProtocolVersions experimentalHardForksEnabled =
    CardanoProtocolVersions {
        protVerByron   = Byron.Update.ProtocolVersion
                           (SL.getVersion (eraProtVerLow @Shelley))
                           0 0
      , protVerShelley = SL.ProtVer (eraProtVerLow @Allegra) 0
      , protVerAllegra = SL.ProtVer (eraProtVerLow @Mary) 0
      , protVerMary    = SL.ProtVer (eraProtVerLow @Alonzo) 0
      , protVerAlonzo  = SL.ProtVer (eraProtVerLow @Babbage) 0
      , protVerBabbage = if experimentalHardForksEnabled then
                           SL.ProtVer (eraProtVerLow @Conway) 0
                         else
                           SL.ProtVer (eraProtVerHigh @Babbage) 0
      , protVerConway  = SL.ProtVer (eraProtVerHigh @Conway) 0
      }
```
But I think it is still useful to see what these numbers are explicitly, so if we do this, we could copy the current definition just below and have a test that asserts that they are the same. WDYT?

Reply:
> we could copy the current definition just below and have a test that asserts that they are the same.

If I recall correctly, ledger defines these values at the type-level. So you could design it to incur a type error instead of a test error when the indirected definition and the concrete definition disagree.